dlib/dnn/layers.h

```diff
@@ -594,9 +594,334 @@ namespace dlib
         >
     using cont = add_layer<cont_<num_filters,nr,nc,stride_y,stride_x>, SUBNET>;
 
+
+// ----------------------------------------------------------------------------------------
+
+    template <
+        int _scale_y,
+        int _scale_x,
+        unsigned char _method
+        >
+    class upsample_
+    {
+    public:
+
+        static_assert(_scale_y > 0, "The scale factor must be > 0");
+        static_assert(_scale_x > 0, "The scale factor must be > 0");
+        static_assert(0 <= _method && _method < 3, "The method must be between 0 and 2");
+
+        upsample_(
+        )  
+        {
+        }
+
+        long scale_y() const { return _scale_y; }
+        long scale_x() const { return _scale_x; }
+        unsigned char method() const { return _method; }
+
+
+        inline point map_output_to_input (
+            point p
+        ) const
+        {
+            p.x() = (p.x()/scale_x());
+            p.y() = (p.y()/scale_y());
+            return p;
+        }
+
+        inline point map_input_to_output (
+            point p
+        ) const
+        {
+            p.x() = p.x()*scale_x();
+            p.y() = p.y()*scale_y();
+            return p;
+        }
+
+        upsample_ (
+            const upsample_& item
+        )  
+        {
+            (void)item;
+            // this->conv is non-copyable and basically stateless, so we have to write our
+            // own copy to avoid trying to copy it and getting an error.
+        }
+
+        upsample_& operator= (
+            const upsample_& item
+        )
+        {
+            if (this == &item)
+                return *this;
+
+            // this->conv is non-copyable and basically stateless, so we have to write our
+            // own copy to avoid trying to copy it and getting an error.
+            return *this;
+        }
+
+        template <typename SUBNET>
+        void setup (const SUBNET& sub)
+        {
+            (void)sub;
+            params.set_size(0);
+        }
+
+        template <typename SUBNET>
+        void forward(const SUBNET& sub, resizable_tensor& output)
+        {
+            unsigned int gnr = _scale_y * sub.get_output().nr();
+            unsigned int gnc = _scale_x * sub.get_output().nc();
+            unsigned int gnsamps = sub.get_output().num_samples();
+            unsigned int gk = sub.get_output().k();
+            output.set_size(gnsamps,gk,gnr,gnc);
+            output = 0;
```

Comment:
Maybe.  I find that most of my cuda kernels are limited more by how much and in what pattern I touch GPU memory, rather than by the particular operations inside the kernel.